### ISSUE TITLE
empty vector when using infinite elements

### ISSUE THREAD

**Issue by BalticPinguin** [libMesh/libmesh#819]
Hi,
in the github-version example miscellaneous-ex1 fails with the message
**error:**
[0]PETSC ERROR: ------------------------------------------------------------------------
[0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation, probably memory access out of range
[0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
[0]PETSC ERROR: or see http://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
[0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
[0]PETSC ERROR: likely location of problem given in stack below
[0]PETSC ERROR: ---------------------  Stack Frames ------------------------------------
[0]PETSC ERROR: Note: The EXACT line numbers in the stack are not available,
[0]PETSC ERROR:       INSTEAD the line number of the start of the function
[0]PETSC ERROR:       is given.
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: Signal received
[0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.6.2, unknown
[0]PETSC ERROR: /data/tm162/bin/libmesh/examples/miscellaneous/miscellaneous_ex1/.libs/lt-example-dbg on a arch-linux2-c-debug named bchlb.molekel.physik.uni-rostock.de by tm162 Wed Feb  3 09:22:40 2016
[0]PETSC ERROR: Configure options --with-scalar-type=complex --download-fblaslapack --download-mpich
[0]PETSC ERROR: #1 User provided function() line 0 in  unknown file
application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
[unset]: aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0

**Tracking the variables that are wrong**
I tracked it down to the vetors Ss and St being empty in line 836 of src/fe/inf_fe.C; hence base_fe->dphidxi and bese_fe->dphidtheta seem to be empty. Finally, they should have been calculated in src/fe/fe.C in the function init_shape_functions() (line 506 ff) where the if-condition is not fullilled. 

**Possible solution**
I must admit, I don't really understand the part where these vectors set (see appended patch); to my understanding, the whole expression should be reverted (as explained in my patch); but maybe some other changes are better.

[fe.patch.txt](https://github.com/libMesh/libmesh/files/115502/fe.patch.txt)

Best,
Huber

**Comment by jwpeterson** [comment]
On Wed, Feb 3, 2016 at 4:03 AM, Hubert Weißmann <notifications@github.com>
wrote:

> Hi,
> in the github-version example miscellaneous-ex1 fails with the message
> *error:*
> [0]PETSC ERROR:
> ------------------------------------------------------------------------
> [0]PETSC ERROR: Caught signal number 11 SEGV: Segmentation Violation,
> probably memory access out of range
> [0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
> [0]PETSC ERROR: or see
> http://wwwmcsanlgov/petsc/documentation/faqhtml#valgrind
> [0]PETSC ERROR: or try http://valgrindorg on GNU/linux and Apple Mac OS X
> to find memory corruption errors
> [0]PETSC ERROR: likely location of problem given in stack below
> [0]PETSC ERROR: --------------------- Stack Frames
> ------------------------------------
> [0]PETSC ERROR: Note: The EXACT line numbers in the stack are not
> available,
> [0]PETSC ERROR: INSTEAD the line number of the start of the function
> [0]PETSC ERROR: is given
> [0]PETSC ERROR: --------------------- Error Message
> --------------------------------------------------------------
> [0]PETSC ERROR: Signal received
> [0]PETSC ERROR: See http://wwwmcsanlgov/petsc/documentation/faqhtml for
> trouble shooting
> [0]PETSC ERROR: Petsc Release Version 362, unknown
> [0]PETSC ERROR:
> /data/tm162/bin/libmesh/examples/miscellaneous/miscellaneous_ex1/libs/lt-example-dbg
> on a arch-linux2-c-debug named bchlbmolekelphysikuni-rostockde by tm162 Wed
> Feb 3 09:22:40 2016
> [0]PETSC ERROR: Configure options --with-scalar-type=complex
> --download-fblaslapack --download-mpich
> [0]PETSC ERROR: #1 <https://github.com/libMesh/libmesh/issues/1> User
> provided function() line 0 in unknown file
> application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
> [unset]: aborting job:
> application called MPI_Abort(MPI_COMM_WORLD, 59) - process 0
>
> *Tracking the variables that are wrong*
> I tracked it down to the vetors Ss and St being empty in line 836 of
> src/fe/inf_feC; hence base_fe->dphidxi and bese_fe->dphidtheta seem to be
> empty Finally, they should have been calculated in src/fe/feC in the
> function init_shape_functions() (line 506 ff) where the if-condition is not
> fullilled
>

It's definitely possible that misc_ex1 + complex numbers has regressed.
Would you mind compiling in debug mode, and running it through a debugger?
That might give us a stack trace and a bit more to work with.


> *Possible solution*
> I must admit, I don't really understand the part where these vectors set
> (see appended patch); to my understanding, the whole expression should be
> reverted (as explained
>
> in my patch); but maybe some other changes are better
>
Looks like your patch was removed by the mailing list, sorry about that.  I
can't recommend strongly enough (and I think Roy also mentioned) getting a
GitHub account and using that to share patches.  It makes it much easier
for others to quickly review and comment on your changes.

-- 
John

**Comment by jwpeterson** [comment]
> Looks like your patch was removed by the mailing list, sorry about that.

Oops, just realized that your patch link is available in the issue.

**Comment by BalticPinguin** [comment]
Hi,
yea, the patch was a bit hidden; sorry :-)
At the moment I get also errors with my patch; so it seems I didn't test enough. Since the reinstallation needs several hours after changes in that part, I will need some time to get clear about that error.
Sorry for that, I will write as soon I am sure about that.

**Comment by roystgnr** [comment]
Yeah, that's the wrong fix, but you've obviously found a real problem.  Could you let us know what options you gave to configure, and what third-party package versions you're using?

**Comment by BalticPinguin** [comment]
I am still confused; with version 0.9.5-rc1 this helps (even if it might be not the best way) and I had it running with the github-version as well.
So the config-options are
_
./configure --prefix=/home/tm162/bin/libmesh \
   --enable-default-comm-world \
   --disable-strict-lgpl   \
   --enable-complex        \
   --enable-perflog        \
   --enable-qhull          \
   --enable-slepc          \
   --enable-petsc          \
   --enable-trilinos       \
   --enable-triangle       \
   --enable-tetgen         \
   --enable-tecplot        \
   --enable-ifem           \
   --disable-second        \
   --enable-everything     >& configs.conf
_
and slepc/petsc are in version 3.6.2; other packages are not included. (I hope it is not compiler-dependent!)

**Comment by roystgnr** [comment]
Thanks.  I'll see if I can replicate now.  Do you have any LIBMESH_RUN or LIBMESH_OPTIONS set?

**Comment by BalticPinguin** [comment]
no, I don't...

**Comment by roystgnr** [comment]
Okay; you shouldn't need to.  Just making sure that I'm not trying to replicate in serial a problem that you encountered only in parallel, or vice-versa.

**Comment by roystgnr** [comment]
And I've had no luck.  I'm seconding John's request; could you run in debug mode and give us a stack trace of the segfault (or of any assertion failure which occurs instead)?

**Comment by roystgnr** [comment]
To clarify: "no luck" because miscellaneous_ex1 runs fine for me with your configure options.  That would be lucky if *I* was the one wanting to mix infinite elements and complex numbers...

**Comment by BalticPinguin** [comment]
Sorry for the long delay; I did recompile the whole library (to avoid some interference with previous installations with other options that I had) and it takes some time; for some reason it even faild with the same script as I used before...
However, using version 0.9.5-rc1 the problem remains. Here, the error + backtrace is:

`
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff3ba42a8 in libMesh::InfFE<3u, (libMesh::FEFamily)12, (libMesh::InfMapType)0>::combine_base_radial(libMesh::Elem const*) ()
   from /home/tm162/bin/libmesh/lib/libmesh_opt.so.0
Missing separate debuginfos, use: debuginfo-install cyrus-sasl-lib-2.1.23-15.el6_6.2.x86_64 glibc-2.12-1.166.el6.x86_64 keyutils-libs-1.4-5.el6.x86_64 krb5-libs-1.10.3-37.el6_6.x86_64 libX11-1.6.0-2.2.el6.x86_64 libXau-1.0.6-4.el6.x86_64 libcom_err-1.41.12-21.el6.x86_64 libcurl-7.19.7-40.el6_6.4.x86_64 libidn-1.18-2.el6.x86_64 libselinux-2.0.94-5.8.el6.x86_64 libssh2-1.4.2-1.el6_6.1.x86_64 libxcb-1.9.1-2.el6.x86_64 nspr-4.10.8-1.el6_6.x86_64 nss-3.18.0-5.3.el6_6.x86_64 nss-softokn-freebl-3.14.3-22.el6_6.x86_64 nss-util-3.18.0-1.el6_6.x86_64 openldap-2.4.39-8.el6.x86_64 openssl-1.0.1e-30.el6.8.x86_64 zlib-1.2.3-29.el6.x86_64
(gdb) bt
#0  0x00007ffff3ba42a8 in libMesh::InfFE<3u, (libMesh::FEFamily)12, (libMesh::InfMapType)0>::combine_base_radial(libMesh::Elem const*) ()
   from /home/tm162/bin/libmesh/lib/libmesh_opt.so.0
#1  0x00007ffff3bc00e3 in libMesh::InfFE<3u, (libMesh::FEFamily)12, (libMesh::InfMapType)0>::reinit(libMesh::Elem const*, std::vector<libMesh::Point, std::allocator<libMesh::Point> > const*, std::vector<double, std::allocator<double> > const*) () from /home/tm162/bin/libmesh/lib/libmesh_opt.so.0
#2  0x0000000000407343 in assemble_wave(libMesh::EquationSystems&, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
    at miscellaneous_ex1.C:376
#3  0x00007ffff3fe1e79 in libMesh::System::assemble() () from /home/tm162/bin/libmesh/lib/libmesh_opt.so.0
#4  0x00007ffff3fc773e in libMesh::LinearImplicitSystem::solve() () from /home/tm162/bin/libmesh/lib/libmesh_opt.so.0
#5  0x000000000040db25 in main () at miscellaneous_ex1.C:193`

**Comment by roystgnr** [comment]
Could you run with METHOD=dbg and see if we get any error more useful than a segfault?

**Comment by jwpeterson** [comment]
You may have already figured out something similar to this, but in the build directory for the example in question, you can run:
```
libtool --mode=execute gdb ./example-dbg
```
(replacing gdb with lldb or your favorite debugger).  Then type:
```
b MPI_Abort
```
(just to guarantee the code pauses before exiting) followed by
```
r
```
and finally
```
bt
```
when it segfaults.

> I hope, this helps you; I will try to get this for the github-version as well.

I wouldn't worry about that -- 0.9.5 and master have not diverged at all as far as infinite elements and complex number support.

**Comment by BalticPinguin** [comment]
Hi, 
thanks for your help, I am not very familiar with these tools (which is a mistake, as I see).
So running as described above, gives me
`Program received signal SIGABRT, Aborted.
0x0000003d39e32625 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install cyrus-sasl-lib-2.1.23-15.el6_6.2.x86_64 glibc-2.12-1.166.el6.x86_64 keyutils-libs-1.4-5.el6.x86_64 krb5-libs-1.10.3-37.el6_6.x86_64 libX11-1.6.0-2.2.el6.x86_64 libXau-1.0.6-4.el6.x86_64 libcom_err-1.41.12-21.el6.x86_64 libcurl-7.19.7-40.el6_6.4.x86_64 libidn-1.18-2.el6.x86_64 libselinux-2.0.94-5.8.el6.x86_64 libssh2-1.4.2-1.el6_6.1.x86_64 libxcb-1.9.1-2.el6.x86_64 nspr-4.10.8-1.el6_6.x86_64 nss-3.18.0-5.3.el6_6.x86_64 nss-softokn-freebl-3.14.3-22.el6_6.x86_64 nss-util-3.18.0-1.el6_6.x86_64 openldap-2.4.39-8.el6.x86_64 openssl-1.0.1e-30.el6.8.x86_64 zlib-1.2.3-29.el6.x86_64
(gdb) bt
#0  0x0000003d39e32625 in raise () from /lib64/libc.so.6
#1  0x0000003d39e33e05 in abort () from /lib64/libc.so.6
#2  0x00007ffff21f41d5 in __gnu_debug::_Error_formatter::_M_error() const () at ../../../../../gcc-5.2.0/libstdc++-v3/src/c++11/debug.cc:782
#3  0x00000000004394fc in std::__debug::vector<std::__debug::vector<double, std::allocator<double> >, std::allocator<std::__debug::vector<double, std::allocator<double> > > >::operator[](unsigned long) const () at /usr/include/c++/4.8.2/debug/vector:353
#4  0x00007ffff68dc34e in libMesh::InfFE<3u, (libMesh::FEFamily)12, (libMesh::InfMapType)0>::combine_base_radial(libMesh::Elem const*) () at src/fe/inf_fe.C:837
#5  0x00007ffff68d8e82 in libMesh::InfFE<3u, (libMesh::FEFamily)12, (libMesh::InfMapType)0>::reinit(libMesh::Elem const*, std::__debug::vector<libMesh::Point, std::allocator<libMesh::Point> > const*, std::__debug::vector<double, std::allocator<double> > const*) () at src/fe/inf_fe.C:221
#6  0x0000000000430a94 in assemble_wave(libMesh::EquationSystems&, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
    at miscellaneous_ex1.C:379
#7  0x00007ffff6f2d762 in libMesh::System::user_assembly() () at src/systems/system.C:1936
#8  0x00007ffff6f26082 in libMesh::System::assemble() () at src/systems/system.C:480
#9  0x00007ffff6f0bdb9 in libMesh::ImplicitSystem::assemble() () at src/systems/implicit_system.C:202
#10 0x00007ffff6dd6b60 in libMesh::LinearImplicitSystem::assemble() () at ./include/libmesh/linear_implicit_system.h:103
#11 0x00007ffff6f16ee5 in libMesh::LinearImplicitSystem::solve() () at src/systems/linear_implicit_system.C:113
#12 0x00000000004302ed in main () at miscellaneous_ex1.C:196
`
But `b MPI_Abort` gives me`Function` "MPI_Abort" not `defined. Moreover, the appearance of /usr/include/c++/... is strange to me; I run it on a cluster where I loaded gcc as a module which is located at /cluster/... Could my errors be due to inconsistent links/library usage instead of code-error?
Thanks for all your help!

**Comment by jwpeterson** [comment]
(FYI, I edited your post to use triple-backticks and make the stack trace easier to read.)

Thanks for the stack trace.  On the line in question (inf_fe.C:837) there are several std::vector accesses, it's difficult to say which one is doing the out-of-range access without doing some more debugging in frame 4. 
```
dphidxi  [ti][bp+rp*n_base_qp] = Ss[bi][bp] * mode[ri][rp] * som[rp];
```


> But b MPI_Abort gives meFunction "MPI_Abort" not `defined.

Interesting, but since you got the stack trace, that step is apparently not necessary.

> Could my errors be due to inconsistent links/library usage instead of code-error?

This can absolutely cause weird segfaults, and they often show up down in STL functions; I often see them in std::string.  Is this a situation where you compiled on the head node and then the compute nodes have a slightly different configuration?  In fact, since Roy cannot replicate the issue on his machine, this is what I suspect the issue is...

**Comment by BalticPinguin** [comment]
Yes, that particular line is what I figured out as well. And the reason I found (by trying to print the sizes) is that Ss (as well as St) just are empty (see my initial post).
At the moment I compile and run the tests both on the head-node. But I will have a chat with our admin how to get a more reliable installation on that machine hoping that it will make my errors disappearing.
Thank you again for your help and your patience.

**Comment by BalticPinguin** [comment]
Sorry for the long delay; I tried different configurations and finally succeeded now with one.
I don't get these errors any more. 
So you can mark it as closed.
Best,
Hubert

**Comment by BalticPinguin** [comment]
One more thing: I have now an other problem with these vectors; should I open an other issue for that or discuss it here?

**Comment by pbauman** [comment]
> I tried different configurations and finally succeeded now with one.
I don't get these errors any more. 

Great, glad to hear it.

> One more thing: I have now an other problem with these vectors; should I open an other issue for that or discuss it here?

If it's a separate issue from this one, it's probably better for a new issue. If it's related to this one, you can post here.

**Comment by jwpeterson** [comment]
Continuing related discussion on #830.
